src/3d/chunks/qgschunkedentity_p.cpp

```diff
@@ -0,0 +1,379 @@
+#include <QVector4D>
+#include "qgs3dutils.h"
+#include "qgschunkedentity_p.h"
+#include "qgschunkboundsentity_p.h"
+#include "qgschunklist_p.h"
+#include "qgschunkloader_p.h"
+#include "qgschunknode_p.h"
+
+///@cond PRIVATE
+
+static float screenSpaceError( float epsilon, float distance, float screenSize, float fov )
+{
+  /* This routine approximately calculates how an error (epsilon) of an object in world coordinates
+   * at given distance (between camera and the object) will look like in screen coordinates.
+   *
+   * the math below simply uses triangle similarity:
+   *
+   *             epsilon                       phi
+   *   -----------------------------  = ----------------
+   *   [ frustum width at distance ]    [ screen width ]
+   *
+   * Then we solve for phi, substituting [frustum width at distance] = 2 * distance * tan(fov / 2)
+   *
+   *  ________xxx__      xxx = real world error (epsilon)
+   *  \     |     /        x = screen space error (phi)
+   *   \    |    /
+   *    \___|_x_/   near plane (screen space)
+   *     \  |  /
+   *      \ | /
+   *       \|/    angle = field of view
+   *       camera
+   */
+  float phi = epsilon * screenSize / ( 2 * distance * tan( fov * M_PI / ( 2 * 180 ) ) );
+  return phi;
+}
+
+static float screenSpaceError( QgsChunkNode *node, const QgsChunkedEntity::SceneState &state )
+{
+  float dist = node->bbox().distanceFromPoint( state.cameraPos );
+
+  // TODO: what to do when distance == 0 ?
+
+  float sse = screenSpaceError( node->error(), dist, state.screenSizePx, state.cameraFov );
+  return sse;
+}
+
+QgsChunkedEntity::QgsChunkedEntity( const QgsAABB &rootBbox, float rootError, float tau, int maxLevel, QgsChunkLoaderFactory *loaderFactory, Qt3DCore::QNode *parent )
+  : Qt3DCore::QEntity( parent )
+  , mNeedsUpdate( false )
+  , mTau( tau )
+  , mMaxLevel( maxLevel )
+  , mChunkLoaderFactory( loaderFactory )
+  , mMaxLoadedChunks( 512 )
+{
+  mRootNode = new QgsChunkNode( 0, 0, 0, rootBbox, rootError );
+  mChunkLoaderQueue = new QgsChunkList;
+  mReplacementQueue = new QgsChunkList;
+}
+
+
+QgsChunkedEntity::~QgsChunkedEntity()
+{
+  // derived classes have to make sure that any pending active job has finished / been canceled
+  // before getting to this destructor - here it would be too late to cancel them
+  // (e.g. objects required for loading/updating have been deleted already)
+  Q_ASSERT( !mActiveJob );
+
+  // clean up any pending load requests
+  while ( !mChunkLoaderQueue->isEmpty() )
+  {
+    QgsChunkListEntry *entry = mChunkLoaderQueue->takeFirst();
+    QgsChunkNode *node = entry->chunk;
+
+    if ( node->state() == QgsChunkNode::QueuedForLoad )
+      node->cancelQueuedForLoad();
+    else if ( node->state() == QgsChunkNode::QueuedForUpdate )
+      node->cancelQueuedForUpdate();
+    else
+      Q_ASSERT( false );  // impossible!
+  }
+
+  delete mChunkLoaderQueue;
```

Comment:
Do you think that a unique_ptr would be interesting in this case?

Reply:
Probably... just being lazy to actively start using new patterns :-)